index.html

```diff
@@ -140,67 +166,250 @@ <h2>
+        badge in the page's <a data-cite=
+        "html#the-title-attribute"><code>title</code></a> or <a data-cite=
+        "html#rel-icon"><code>icon</code></a>.
+      </p>
+      <p class="note">
+        These methods return a {{Promise}}, even though there are no
+        asynchronous failure conditions (the only failure conditions are WebIDL
+        type errors). This is to allow us to add asynchronous failure
+        conditions in the future without altering how type errors are returned.
       </p>
       <section>
         <h3>
-          <dfn>set()</dfn> function
+          <dfn data-dfn-for="Navigator">setClientBadge()</dfn> method
         </h3>
         <p>
-          When the <code>set()</code> function is called with argument
+          When the <code>setClientBadge()</code> method is called with argument
           <var>contents</var>, run the following steps:
         </p>
         <ol>
-          <li>If <var>contents</var> is omitted, set the application's badge to
-          <a>flag</a>.
+          <li>Let <var>document</var> be the <a data-cite=
+          "html#current-settings-object">current settings object</a>'s
+          <a data-cite="html#responsible-document">responsible document</a>.
+          </li>
+          <li>If <var>contents</var> is omitted, set the badge associated with
+          <var>document</var> to <a>flag</a>.
           </li>
-          <li>If <var>contents</var> is 0, set the application's badge to
-          <a>nothing</a>.
+          <li>Else, if <var>contents</var> is 0, set the badge associated with
+          <var>document</var> to <a>nothing</a>.
           </li>
-          <li>Else, set the application's badge to <var>contents</var>.
+          <li>Else, set the badge associated with <var>document</var> to <var>
+            contents</var>.
+          </li>
+          <li>Return a <a data-cite="webidl#a-promise-resolved-with">resolved
+          promise</a>.
+          </li>
+        </ol>
+      </section>
+      <section>
+        <h3>
+          <dfn data-dfn-for="Navigator">clearClientBadge()</dfn> method
+        </h3>
+        <p>
+          When the <code>clearClientBadge()</code> method is called, set the
+          the badge associated with the <a data-cite=
+          "html#current-settings-object">current settings object</a>'s
+          <a data-cite="html#responsible-document">responsible document</a> to
+          <a>nothing</a>, and return a <a data-cite=
+          "webidl#a-promise-resolved-with">resolved promise</a>.
+        </p>
+      </section>
+      <section>
+        <h3>
+          <dfn data-dfn-for="NavigatorBadge">setAppBadge()</dfn> method
+        </h3>
+        <p>
+          When the <code>setAppBadge()</code> method is called with argument
+          <var>contents</var>, run the following steps:
+        </p>
+        <ol>
+          <li>Let <var>result</var> be the result of <a>determining the
```

Comment:
Can "determining the matching applications" be done without IPC? To be safe, return a promise and continue in parallel.

Reply:
It's been changed so the error detection happens in the xxxAppBadge methods. I don't think this needs to return a promise. Even if it requires IPC, that's fine because it takes place in the "in parallel" steps of those algorithms, so you can take as much time as you need to execute it.